README
INTRODUCTION:

Inline::Java - Write Perl classes in Java.

Inline::Java lets you write Perl classes in Java.

Example:

    use Inline Java => <<'END';
        class JAxH {
            public JAxH(String x){
                System.out.println("Just Another " + x + " Hacker") ;
            }
        }
    END
    
    new JAxH('Inline') ;

When run, this complete program prints:

    Just Another Inline Hacker


-------------------------------------------------------------------------------
JNI (JAVA NATIVE INTERFACE) EXTENSION:

Inline::Java now provides a JNI extension that allows you to load the Java 
virtual machine as shared object instead of running it as a separate process. 

See README.JNI for more information on building the JNI extension.


-------------------------------------------------------------------------------
INSTALLATION:

- This module requires Inline version 0.44 or higher to be installed.
- It also requires a version of the Java SDK 1.2 or higher to be 
  installed. You can get the latest Java SDK from Sun Microsystems
  at http://java.sun.com. Follow the provided instructions in order
  to install the Java SDK properly.

To install Inline::Java do this:

  % perl Makefile.PL J2SDK=/your/java/dir   (see Note 1)
  % make                                    (see Note 2)
  % make test                               (see Note 3, 4)
  % make install

You have to 'make install' before you can run it successfully. 

Note 1: Under Win95/98/Me, you may need to do 'cd ..' to get back to the
your original directory after the command has completed. Also, you may set
either the JAVA_HOME or the PERL_INLINE_JAVA_J2SDK environment variable to 
/your/java/dir instead of using the J2SDK Makefile.PL parameter.

Note 2: Use nmake on Win32.

Note 3: If you have built the JNI extension and want the test suite to use 
it, you will need to set the PERL_INLINE_JAVA_JNI environment variable to 1 
BEFORE running 'make test'.

Note 4: When testing Inline::Java, it's always a good idea to run 'make test'
twice. The first time you test the building and loading of a module, the 
second time you test loading of an already built module.


-------------------------------------------------------------------------------
FEATURES FOR THIS VERSION:

Inline::Java version 0.52 is a minor upgrade that includes:
    - Fixed JNI on cygwin (many thanks to Eric Rybski for the patch)
    - Improved installation. 'make java' is now performed automatically.
    - Fixed problems with disappearing exceptions by localizing $@.
    - Other minor bug fixes.

See CHANGES for a full change list.

-------------------------------------------------------------------------------
INFORMATION:

= For more information on Inline::Java, see 'perldoc Inline::Java'.
= For information about Inline, see 'perldoc Inline'.
= For information on using Java, visit http://java.sun.org.

Inline::Java's mailing list is inline@perl.org. 
To subscribe, send an email to inline-subscribe@perl.org.

Inline::Java's home page is http://inline.perl.org/java/.

Please send questions and comments to Patrick LeBoutillier <patl@cpan.org>.

Copyright (c) 2001-2005, Patrick LeBoutillier. All Rights Reserved.  


CHANGES
Revision history for Perl extension Inline::Java
------------------------------------------------
0.52  Sun Dec 17 20:46:51 EST 2006
    - Fixed JNI on cygwin (many thanks to Eric Rybski for the patch)
    - Improved installation. 'make java' is now performed automatically.
    - Fixed problems with disappearing exceptions by localizing $@.
    - Other minor bug fixes

0.51  Tue May 23 20:40:07 EDT 2006
    - Several major speed optimizations.
    - Introduction of support for I/O mapping between Perl and Java (Inline::Java::Handle)
    - Applied patches by Andrew Bruno and Tim Bunce for MAC OSX
    - JNI fix for system property passing
        (thanks to Brian Gugliemetti and Jason Stelzer)
    - Added NATIVE_DOUBLES configuration option to avoid loss of precision
      when passing double values between Perl and Java
    - New interface for processing callbacks from java to perl.
    - Added support for java.lang.CharSequence as a primitive type. Any
      Perl scalar passed as a java.lang.CharSequence will instantiate
      a java.lang.String on the Java side
    - Added BUILD_JNI, BUILD_PERL_NATIVES, BUILD_PERL_INTERPRETER and JVM_LIB_TYPE build 
      configuration options to help with automated builds
    - Added BIND configuration option. See docs for details. 
      Note: The Java JVM server noe listens on localhost by default (instead of 0.0.0.0).
    - Other minor bug fixes

0.50  Mon Jan 31 20:14:43 EST 2005
    - Added HOST configuration option to access JVM server remotely
    - Fixed bug with paths in Cygwin
    - Fixed bug with regexp that was badly interpreted under the debugger
    - Extended InlineJavaClassLoader from current thread ClassLoader to enable it to work
      properly under Tomcat
    - Include patch by Ihab Awad that allows running the JVM server inside Tomcat
    - Fixed bug with longs over 32 bits in length and loosened boundary checking in numeric values
      since it was not portable
    - Fixed bug (found by Dean Thayer) with socket not closed properly in Java
    - Other minor bug fixes

0.49  
    - Added PerlInterpreter: require/eval Perl code directly from Java
    - Reworked type casting: changes are NOT backwards compatible :(
    - Callbacks can now be called in list context using "@function_name"
    - Renamed PerlNatives stuff
    - Added external command line tool to start/stop a SHARED_JVM server
    - Applied JNI memory leak patch by Jeff Janes
    - Removed exports from Inline::Java::Portable
    - Split and updated documentation

0.47  Sat Feb 14 10:00:00 EST 2004
    - Fixed bugs in portability code and added HPUX, AIX and Solaris specifics
    - Tweaked CLASSPATH directory order
	- Changed port numbers for SHARED_JVM tests in order to avoid clashes
      with running installations
    - Made PerlNatives optional
    
0.46  Web Feb 04 20:00:00 EST 2004
    - Fixed Natives.xs to work with ExtUtils::ParseXS

0.45  Fri Jan 30 20:00:00 EST 2004
    - Fixed Makefile.PL arguments that were getting lost
    - Fixed deprecated require call
    - Fixed support for symbolic links in J2SDK directory
    - Basic support for J2SDK 1.5
    - Added new experimental feature: PerlNatives

0.44  Sun Nov 23 15:47:06 EST 2003
    - Callbacks from multiple threads are now supported
    - Refactored (again...) studying/.jdat/cache stuff
    - Added PRIVATE mode for use with SHARED_JVM
    - Added DEBUGGER mode that launches jdb
    - Fixed memory leak in JNI code (patch submitted by Dave Blob)

0.43  Tue Oct 14 13:18:25 EDT 2003
    - Restored $VERSION in each .pm file
    - Inline::Java now formerly requires Perl 5.6

0.42  Fri Sep  5 13:18:25 EDT 2003
    - Fixed more CLASSPATH issues. CLASSPATH now works like this:
      * CLASSPATH environment variable is global
      * CLASSPATH configuration option is local to the user class loader
    - Added method cache to increase performance and decrease reflection 
      API calls
    - Altered and documented the study_classes() function
    - Added EXTRA_JAVA_ARGS and EXTRA_JAVAC_ARGS config options

0.41  Mon Jul 14 13:18:25 EDT 2003
    - Fixed CLASSPATH bug
    - Possibly (!) fixed test suite problems under heavy load

0.40  Fri Apr 11 11:00:00 EST 2003
    - Patch by Doug MacEachern to allow running under servlet engines.
    - Patch by John Kinsley for handling empty arrays.
    - Custom ClassLoader allows for better isolation of user code and
      dynamic 'CLASSPATH' (multiple sections/clients are now totally
      independant).
    - Core Java code is now compiled separately (at install time) instead
      of with the user code (for each script). This dramatically improves
      build time.
    - Default port number changed to 0 (next available port number). 
      This allows many multiple users to run concurrently without 
      clashes.
      Note: The default port on systems where this feature is not 
            available has remained 7890.
      Note: The default port for SHARED_JVM mode is now 7891.
    - SHARED_JVM mode now calls release_JVM() automatically.
    - UTF8 support
    - User code can now be a public class and (optionally) inside a package.
    - Callback classes have changed.
      Note: These classes are now in the org.perl.inline.java package.
      Note: PerlException has been renamed InlineJavaPerlException.
      Note: If you presently use callbacks, your code may now no longer compile.
            You will need to apply the changes listed above to fix it.
    - study_classes now returns the package in which the studied classes
	  have been put.
    - Many bug fixes and documentation corrections.

0.33  Mon Jun 17 13:50:14 EDT 2002
    - Improved and more flexible debug output
    - Auto loading of thread libraries under Solaris
	- Basic support for J2SDK 1.4
    - Other minor bug fixes

0.32  Sat Apr  6 11:45:06 EST 2002
    - Restored compatibility with perl 5.005_03
    - Other minor bug fixes

0.31  Mon Feb  4 15:45:06 EDT 2002

Makefile.PL
use ExtUtils::MakeMaker ;

use strict ;
use File::Spec ;
use Cwd ;
use Config ;

require "Java/Portable.pm" ;

print "\nWelcome to the Inline::Java installation procedure.\n\n" ;

# Grab the J2SDK argument
my $jdk_dir = '' ;
for (my $i = 0 ; $i < scalar(@ARGV) ; $i++){
	my $remove = 0 ;
	if ($ARGV[$i] =~ /^J2SDK=(.+)$/){
		$jdk_dir = $1 ;
		$remove = 1 ;
	}
	elsif ($ARGV[$i] =~ /^BUILD_JNI=(.+)$/){
		$main::build_jni = $1 ;
		$remove = 1 ;
	}
	elsif ($ARGV[$i] =~ /^BUILD_PERL_NATIVES=(.+)$/){
		$main::build_perl_natives = $1 ;
		$remove = 1 ;
	}
	elsif ($ARGV[$i] =~ /^BUILD_PERL_INTERPRETER=(.+)$/){
		$main::build_perl_interpreter = $1 ;
		$remove = 1 ;
	}
	elsif ($ARGV[$i] =~ /^JVM_LIB_TYPE=(.+)$/){
		$main::jvm_lib_type = $1 ;
		$remove = 1 ;
	}
	if ($remove){
		splice(@ARGV, $i, 1) ;
		$i-- ;
	}
}

if (! $jdk_dir){
	my $try = $ENV{PERL_INLINE_JAVA_J2SDK} || $ENV{JAVA_HOME} 
		|| Inline::Java::Portable::portable('DEFAULT_J2SDK_DIR') ;
	print "Using $try as J2SDK directory.\n\n" if $try ;
	$jdk_dir = $try ;
}

if (! $jdk_dir){
	my $def_pl = File::Spec->catfile('Java', 'default_j2sdk.pl') ;
	if (-e $def_pl){
		require File::Spec->catfile('Java', 'default_j2sdk.pl') ;
		$jdk_dir = Inline::Java::get_default_j2sdk() ;
	}
	else {
		print <<NO_J2SDK;
A Java 2 SDK is required to install and use Inline::Java. Please 
specify your Java 2 SDK installation directory using the J2SDK 
option to Makefile.PL as such:

    perl Makefile.PL J2SDK=/path/to/your/j2sdk/installation
NO_J2SDK
		exit(1) ;
	}
}
elsif (! -d $jdk_dir){
	print <<BAD_J2SDK;
Java 2 SDK installation directory '$jdk_dir' does not exist.
BAD_J2SDK
	exit(1) ;
}
my $perl_jdk_dir = $jdk_dir ;
$perl_jdk_dir =~ s/'/\'/g ;

# Check directory
my $jdk_bin = Inline::Java::Portable::portable("J2SDK_BIN") ;
my $ext = Inline::Java::Portable::portable('EXE_EXTENSION') ;
foreach my $f ('javac', 'jar', 'java'){
	if (! -f File::Spec->catfile($jdk_dir, $jdk_bin, $f . $ext)){
		my $bf = File::Spec->catfile($jdk_bin, $f . $ext) ;
		print "Can't locate file '$bf' anywhere under '$jdk_dir'\n" ;
	}
}


# Now we have the J2SDK directory and it exists.
# We will create the default_j2sdk.pl file that
# will contain that value for future use.
my $def_jdk = File::Spec->catfile('Java', 'default_j2sdk.pl') ;
open(J2SDK, ">$def_jdk") or
	die("Can't open '$def_jdk' for writing: $!") ;
print J2SDK <<J2SDK_PL;
# This file is created by the Makefile.PL for Inline::Java
# You can modify it if you wish
use strict ;

# The default J2SDK to use for Inline::Java. You can change
# it if this value becomes invalid.
sub Inline::Java::get_default_j2sdk {
	return '$perl_jdk_dir' ;
}
1 ;


J2SDK_PL
close(J2SDK) ;

print <<SAVE_J2SDK;
Default J2SDK for Inline::Java will be '$jdk_dir'.
See module documentation for information on how to use a different J2SDK
or change this default value.

SAVE_J2SDK


# We will now add the building of our Java files to the Makefile.
my $javac = File::Spec->catfile($jdk_dir, $jdk_bin, 'javac' . $ext) ;
my $jar = File::Spec->catfile($jdk_dir, $jdk_bin, 'jar' . $ext) ;                      
my $src_dir = File::Spec->catdir('Java', 'sources', 'org', 'perl', 'inline', 'java') ;
my $src = File::Spec->catfile($src_dir, '*.java') ;
my $obj_dir = File::Spec->catdir('Java', 'classes') ;
my $server_arch = File::Spec->catfile('Java', 'InlineJavaServer.jar') ;
my $user_arch = File::Spec->catfile('Java', 'InlineJavaUser.jar') ;

# Create the object diretory because later we need to put the properties
# file inside it.
if (! -e $obj_dir){
	mkdir($obj_dir) or
		die("Can't create object directory '$obj_dir': $!") ;
}

sub MY::top_targets {
	my $this = shift ;

	my $make = <<MAKE ;
# Added by Inline::Java installation
pure_all :: java
MAKE

	return $make . $this->MM::top_targets() ;
}


my $INSTALLSITEARCH = '' ;
my $INST_ARCHLIB = '' ;
sub MY::postamble {
	my $this = shift ;

	my $java_src = join(' ', glob($src), File::Spec->catfile($obj_dir, 'InlineJava.properties')) ;

	my $make = <<MAKE ;
# Added by Inline::Java installation
JAVA_SRC=$java_src 

java.ts: \$(JAVA_SRC)
	\@\$(MKPATH) $obj_dir
	"$javac" -deprecation -g -d $obj_dir $src
	"$jar" cf $server_arch -C $obj_dir org -C $obj_dir InlineJava.properties
	"$jar" cf $user_arch -C $obj_dir InlineJavaUserClassLink.class
	\@\$(TOUCH) java.ts

java :: java.ts
MAKE

	# Used for PerlNatives
	$INSTALLSITEARCH = expand_macros($this, 'INSTALLSITEARCH') ;
	$INST_ARCHLIB = expand_macros($this, 'INST_ARCHLIB') ;

	return $make ;
}


sub expand_macros { 
	my $mm = shift ;
	my $var = shift ;

	my $val = $mm->{$var} ;
	while ($val =~ s/\$\((.*?)\)/$mm->{$1}/){}
	$val =~ s/\\/\\\\/g ;

	return $val ;
}


# Write the Makefile
my $natives_test = File::Spec->catdir('Java', 'Natives', '_Inline_test') ;
my $perlinterp_test = File::Spec->catdir('Java', 'PerlInterpreter', '_Inline_test') ;
WriteMakefile(
	NAME => 'Inline::Java',
	VERSION_FROM => 'Java.pm',
	DIR => ['Java'],
	PREREQ_PM => {
		Inline	=> 0.44,
		Test => 1.13,
		'MIME::Base64' => 0,
	},
	PM => {
		'Java.pm' => File::Spec->catfile('$(INST_LIBDIR)', 'Java.pm'),
		'Java.pod'=> File::Spec->catfile('$(INST_LIBDIR)', 'Java.pod'),
		$server_arch => File::Spec->catfile('$(INST_LIBDIR)', $server_arch),
		$user_arch => File::Spec->catfile('$(INST_LIBDIR)', $user_arch),
	},
	clean => {FILES => "$def_jdk _Inline_test $natives_test $perlinterp_test $obj_dir $server_arch $user_arch java.ts"},
) ;


# Add the so_dirs to the default_j2sdk.pl file.
open(J2SDK, ">>$def_jdk") or
	die("Can't open '$def_jdk' for appending: $!") ;
print J2SDK <<J2SDK_PL;
sub Inline::Java::get_default_j2sdk_so_dirs {
	return (
J2SDK_PL
foreach my $d (@main::SO_DIRS){
	$d =~ s/'/\'/g ;
	print J2SDK "\t\t'$d',\n" ;
}
print J2SDK <<J2SDK_PL;
	) ;
}


1 ;
J2SDK_PL
close(J2SDK) ;



# Create the properties that will be included in the jar.
my @perlnatives_so_parts = ("auto", "Inline", "Java", "PerlNatives", 
	"PerlNatives." . Inline::Java::Portable::portable('SO_EXT')) ;
my $install_perlnatives_so = File::Spec->catfile($INSTALLSITEARCH, @perlnatives_so_parts) ;
$install_perlnatives_so = Inline::Java::Portable::portable("SUB_FIX_JAVA_PATH", $install_perlnatives_so) ;
$install_perlnatives_so =~ s/\\/\\\\/g ;
my $test_perlnatives_so = File::Spec->rel2abs(File::Spec->catfile($INST_ARCHLIB, @perlnatives_so_parts)) ;
$test_perlnatives_so = Inline::Java::Portable::portable("SUB_FIX_JAVA_PATH", $test_perlnatives_so) ;
$test_perlnatives_so =~ s/\\/\\\\/g ;

my @perlinterpreter_so_parts = ("auto", "Inline", "Java", "PerlInterpreter",
    "PerlInterpreter." . Inline::Java::Portable::portable('SO_EXT')) ;
my $install_perlinterpreter_so = File::Spec->catfile($INSTALLSITEARCH, @perlinterpreter_so_parts) ;
$install_perlinterpreter_so = Inline::Java::Portable::portable("SUB_FIX_JAVA_PATH", $install_perlinterpreter_so) ;
$install_perlinterpreter_so =~ s/\\/\\\\/g ;
my $test_perlinterpreter_so = File::Spec->rel2abs(File::Spec->catfile($INST_ARCHLIB, @perlinterpreter_so_parts)) ;
$test_perlinterpreter_so = Inline::Java::Portable::portable("SUB_FIX_JAVA_PATH", $test_perlinterpreter_so) ;
$test_perlinterpreter_so =~ s/\\/\\\\/g ;

my $libperl = $Config{libperl} ;
my $dlext = $Config{dlext} ;
my $libperl_so = '' ;
if ($libperl =~ /\.$dlext$/){
	$libperl_so = File::Spec->catfile($Config{installarchlib}, 'CORE', $libperl) ;
}

my $prop = File::Spec->catfile($obj_dir, 'InlineJava.properties') ;
open(PROP, ">$prop") or
	die("Can't open '$prop' for writing: $!") ;
print PROP <<PROP;
# This file is created by the Makefile.PL for Inline::Java
inline_java_perlnatives_so_install = $install_perlnatives_so
inline_java_perlnatives_so_test    = $test_perlnatives_so
inline_java_perlinterpreter_so_install = $install_perlinterpreter_so
inline_java_perlinterpreter_so_test    = $test_perlinterpreter_so
inline_java_libperl_so	= $libperl_so
PROP
close(PROP) ;

# Clean up the Makefile for Win95/98/Me
if (Inline::Java::Portable::portable('COMMAND_COM')){
	print "\nFixing Makefile for Win95/98/Me...\n" ;
	open(MAKEFILE, "<Makefile") or die "Can't open Makefile for reading" ;
	my @lines = <MAKEFILE> ;
	close(MAKEFILE) ;
	open(MAKEFILE, ">Makefile") or die "Can't open Makefile for writing" ;
	foreach my $line (@lines){
		if ($line !~ /^\s*((\@\[)|(\]))\s*$/){
			print MAKEFILE $line ;
		}
	}
	close(MAKEFILE) ;
}

my $make = Inline::Java::Portable::portable('MAKE') ;
print "\nYou can continue the installation with the following commands:\n" ;
print "  % $make\n" ;
print "  % $make test\n" ;
print "  % $make install\n" ; 
